waku-org/nwaku
# chore: Postgres migrations

## Description
Adding the needed logic so that we can safely change the _Postgres_ database schema.
The approach is different from the current migration logic in _SQLite_. In this case, we avoid the use of external resources and instead, use a compiled `consts` that contains the migration scripts.
Now I'm setting the current schema version to 1. Notice that this PR includes a migration script aimed at converting the current table into a partitioned table, and this will happen in the upcoming PR, where the partition logic is implemented. Therefore, even though we have the migration script for version 2, we are still at version 1.

## Changes

- New `breakIntoStatements` proc in `postgres_driver/migrations.nim` to support handling PL/SQL scripts.
- Refactor Postgres unit tests. Now the table is created by the migration procedures.
- Starting to use a new "VERSION" table, where we simply store the current schema version.

## How to test

1. Start node A with `./build/wakunode2 --config-file=cfg_node_a.txt`
[cfg_node_a.txt](https://github.com/waku-org/nwaku/files/14392566/cfg_node_a.txt)

2. In a separate terminal, start a Postgres database locally, with `docker compose -f postgres-docker-compose.yml up`:
Content of postgres-docker-compose.yml:
```YML
version: "3.8"

services:
  db:
    image: postgres:15.4-alpine3.18
    #image: hartmutcouk/pg-repack-docker:1.4.8 # https://github.com/hartmut-co-uk/pg-repack-docker/blob/master/README.md
    ## if we use the 'pg-repack-docker' image, we will need to call the next command externally:
    ## PGPASSWORD=test123 ~/utils/pg_repack-1.4.8/bin/pg_repack -U postgres -h 127.0.0.1 -p 5432 -d postgres --table messages
    ## With that, the database size gets reduced.
    restart: always
    volumes:
      - ./postgres_cfg/postgresql.conf:/etc/postgresql/postgresql.conf
      - ./postgres_cfg/db.sql:/docker-entrypoint-initdb.d/db.sql
      - ./postgres-data:/var/lib/postgresql/data:Z
    command: postgres -c config_file=/etc/postgresql/postgresql.conf
    environment:
      POSTGRES_PASSWORD: test123
    ports:
      - "5432:5432"
```
Notice that the following lines might not be needed. I cannot push the files. You can use the default settings:
      - ./postgres_cfg/postgresql.conf:/etc/postgresql/postgresql.conf
      - ./postgres_cfg/db.sql:/docker-entrypoint-initdb.d/db.sql

3. Start node B with `./build/wakunode2 --config-file=cfg_node_b.txt`
[cfg_node_b.txt](https://github.com/waku-org/nwaku/files/14392567/cfg_node_b.txt)

4. Publish a few messages during some time with `bash msg_publisher.sh`

Being `msg_publisher.sh`:
```bash
#!/bin/sh

while true
do
  payload_size=76800 
  payload=$(openssl rand -hex ${payload_size})
  
  # Example on how to publish to autosharding endpoint
  echo '{"payload":"'${payload}'","contentTopic":"my-ctopic-1", "timestamp":'$(date +%s%N)'}' > msg.txt
  curl -X POST "http://127.0.0.1:8645/relay/v1/messages/%2Fwaku%2F2%2Fdefault-waku%2Fproto"  -H "content-type: application/json"  -d @msg.txt
  
  sleep 0.1
done
```
5. Stop script started in 3 after few minutes
6. Stop node B
7. Change the target version from 1 to 2 in `waku/waku_archive/driver/postgres_driver/migrations.nim`
8. Recompile `wakunode2` and run B again.
9. Without sending any message (the partition logic will come soon,) access directly the database (with Postbird f.e.) and there you should see that the `messages` table is a partitioned table and it exists a partition that contains all the messages stored in the previous raw `messages` table.
![image](https://github.com/waku-org/nwaku/assets/128452529/2b119ddb-281e-4b62-a625-e0d62fa5eca3)

## Issue

closes https://github.com/waku-org/nwaku/issues/2249
